### ISSUE TITLE
Validation Data is not generated properly for my dataset

### ISSUE THREAD

**Issue by RaviKumarAndroid** [jdb78/pytorch-forecasting#1]
Hi, Really appreciate your work on the TFT. 

I am trying to use my own dataset in the code but there seems to be a bug  due to which the dataset is not being loaded properly for validation.
The train dataloader is good. but the validation dataloader only has one batch and also validation(TimeSeriesDataSet) has only 1 entry. 

Below is my complete code
```
data = load_csv()
data['date']= pd.to_datetime(data['date'])
data.reset_index(inplace=True, drop=True)
data.reset_index(inplace=True)
data.rename(columns={'index':'time_idx'}, inplace=True) # I use index as time_idx since my data is of minute frequency

validation_len = int(len(data) * 0.1)
training_cutoff = int(len(data)) - validation_len

max_encode_length = 36
max_prediction_length = 6

print('Len of training data is : ',len(data[:training_cutoff]))
print('Len of val data is : ',len(data[training_cutoff:]))
training = TimeSeriesDataSet(
    data[:training_cutoff],
    time_idx="time_idx",
    target="T",
    group_ids=["Symbol"],
    max_encoder_length=max_encode_length,
    max_prediction_length=max_prediction_length,
    static_categoricals=["Symbol"],
    static_reals=[],
    time_varying_known_categoricals=[
        "hour_of_day",
        "day_of_week",
    ],
    time_varying_known_reals=[
        "time_idx",
    ],
    time_varying_unknown_categoricals=[],
    time_varying_unknown_reals=["V1", "V2","V3", "T", "V4"],
    constant_fill_strategy={"T": 0},
    dropout_categoricals=[],
)
print('Max Prediction Index : ',training.index.time.max())
validation = TimeSeriesDataSet.from_dataset(training, data, min_prediction_idx=training.index.time.max()+1)
batch_size = 128
train_dataloader = training.to_dataloader(train=True, batch_size=batch_size, num_workers=1)
val_dataloader = validation.to_dataloader(train=False, batch_size=batch_size, num_workers=1)

print(len(training), len(validation))
print(len(train_dataloader), len(val_dataloader))`
```

This is what gets printed by the code : 
```
Len of training data is : 25920
Len of val data is : 2880
Min Prediction Index Value is 25919

25920 1
202 1
```

You can see that the training dataset is good and the batches are also okay but validation batch is also 1 and dataset length is also 1. 

**One more thing. if i use predict = False it generates validation data correctly but another bug arises due to that.**
**if i use predict = True, only 1 batch and 1 sequence is given**
**if i use predict_mode = true on the training dataset it also generates only 1 batch.**

Here is a sample of my CSV
[sample_data.csv.zip](https://github.com/jdb78/pytorch_forecasting/files/4977518/sample_data.csv.zip)

Please Help

**Comment by RaviKumarAndroid** [comment]
Please help me how to use the attached dataset through the forecasting library.

**Comment by jdb78** [comment]
The first error is due to an unfortunate default in the `from_dataset` method. The default was `predict=True` which causes to the dataset to select the last samples per timeseries for prediction - as you only have one timeseries there is only one sample). I changed the default to `predict=False`. You might want to pass now `stop_randomization=True` for the validation dataset.

The second error is a genuine error (1 off error) which I just fixed. Thanks for bringing this to my attention! The fix is pushed to master, so installing from git should do the job.

**Comment by RaviKumarAndroid** [comment]
Got another error after predict=False

```
File "/Users/wilson/PycharmProjects/stocky/predictors/temporal_fusion_transformer/main_torch_forecasting_stock.py", line 154, in <module>
    tft, train_dataloader=train_dataloader, val_dataloaders=val_dataloader
  File "/usr/local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1044, in fit
    results = self.run_pretrain_routine(model)
  File "/usr/local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1196, in run_pretrain_routine
    False)
  File "/usr/local/lib/python3.7/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 293, in _evaluate
    output = self.evaluation_forward(model, batch, batch_idx, dataloader_idx, test_mode)
  File "/usr/local/lib/python3.7/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 470, in evaluation_forward
    output = model.validation_step(*args)
  File "/Users/wilson/PycharmProjects/stocky/pytorch_forecasting/models/base_model.py", line 126, in validation_step
    log, _ = self.step(x, y, batch_idx, label="val")
  File "/Users/wilson/PycharmProjects/stocky/pytorch_forecasting/models/temporal_fusion_transformer/__init__.py", line 544, in step
    attention_prediction_horizon=0,  # attention only for first prediction horizon
  File "/Users/wilson/PycharmProjects/stocky/pytorch_forecasting/models/temporal_fusion_transformer/__init__.py", line 576, in interpret_output
    encoder_length_histogram = integer_histogram(out["encoder_lengths"], min=0, max=self.hparams.max_encoder_length)
  File "/Users/wilson/PycharmProjects/stocky/pytorch_forecasting/utils.py", line 27, in integer_histogram
    dim=0, index=uniques - min, src=counts
RuntimeError: index 37 is out of bounds for dimension 0 with size 37
```

**Comment by jdb78** [comment]
This is probably due to only having a single timeseries in the dataset. I will have a closer look later or on the weekend. You could probably run with `log_interval=-1` in the meantime and skip the interpretation step alltogether.

**Comment by RaviKumarAndroid** [comment]
Yeah, For Now I commented out just the interpretation part because i would like to be able to see prediction outputs in tensorboard. But I found an issue in the visualisation. 
**You are plotting Non Normalised target vs output(not yet denormalised).**

<img width="717" alt="Screenshot 2020-07-29 at 1 24 55 AM" src="https://user-images.githubusercontent.com/21041666/88714375-5bff7c80-d13a-11ea-8a1a-d4c71061a437.png">
See the scale of the graph. The target value is about 300 and prediction is about 0 point something. So its not denormalised back again

**Comment by jdb78** [comment]
Sorry for taking a bit of time. The plotting functionality is not broken. However, I think there are two issues here:

1. The network has probably not trained for long enough
2. To train, timeseries should be normalized (I think almost all networks require this) for faster convergence and better generalisation.

I am working on this normalization so you do not have to take care of it manually.

**Comment by RaviKumarAndroid** [comment]
Thanks, I will try the model again once the normalization part is done.

**Comment by RaviKumarAndroid** [comment]
Also please try to fix the issue : https://github.com/jdb78/pytorch_forecasting/issues/1#issuecomment-664624574

**Comment by jdb78** [comment]
Give it another try. I fixed a couple of issues including target normalization. Keep in mind that the strength of the algorithm is forecasting multiple timeseries.

**Comment by jdb78** [comment]
Closing this as there were numerous fixed to the package - thanks again for reporting this issue! Please feel encouraged to raise a new issue in case you encounter any bugs.
